src/cargo/core/resolver/activation_key.rs

```diff
@@ -0,0 +1,73 @@
+use std::collections::HashSet;
+use std::hash::{Hash, Hasher};
+use std::num::NonZeroU64;
+use std::sync::{Mutex, OnceLock};
+
+use crate::core::SourceId;
+use crate::util::interning::InternedString;
+
+static ACTIVATION_KEY_CACHE: OnceLock<Mutex<HashSet<&'static ActivationKeyInner>>> =
+    OnceLock::new();
+
+type ActivationKeyInner = (InternedString, SourceId, SemverCompatibility);
+
+/// The activated version of a crate is based on the name, source, and semver compatibility.
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct ActivationKey {
+    inner: &'static ActivationKeyInner,
+}
+
+impl From<ActivationKeyInner> for ActivationKey {
+    fn from(inner: ActivationKeyInner) -> Self {
+        let mut cache = ACTIVATION_KEY_CACHE
```

Comment:
Note for self: This lock/cash is not a problem for parallelism because at the moment it's only called in `PackageId::new` which is in its own lock/cash.